### PR TITLE
[adapters] Always start a transaction before a step.

### PR DESCRIPTION
**Pull request by ryzhyk**
Need @blp 's help to get this to work with multihost.

----

The controller used to run the circuit in two different modes:

- When a transaction is explicitly requested via an API call
  or by a connector, the controller called `dbsp.start_transaction`
  followed by multiple calls to `dbsp.step()`.

- When no transaction was requested, the controller called
  `dbsp.transaction()`.

In the latter case, the circuit could internally perform a large number of
steps, taking a long time, before returning control to the controller. During
this time, the user didn't get any indication of progress, and wasn't able to
download a support bundle.

This commit addresses the issue by starting a transaction and instantly
initiating transaction commit before performing a step if the transaction was
not explicitly requested. This way, potentially long transactions are broken
down into short steps. The user can monitor progress in the WebConsole just
like during a normal transaction.

### CODE REVIEW

**Review by blp — APPROVED** [review]

File: crates/adapters/src/controller.rs [hunk 1]
```diff
@@ -7146,7 +7180,30 @@ impl ControllerInner {
...
+                            processed_records: self
+                                .status
+                                .global_metrics
+                                .num_total_processed_records(),
+                        };
+                        Some(AdvanceTransaction::StartAndCommit)
+                    }
                 }
             }
         };
```
Comment (mythical-fred) [inline crates/adapters/src/controller.rs:7209]: This PR introduces several significant behavior changes — auto-initiated transactions on every step, snapshot updates restricted to transaction boundaries, API transactions now allowed while the previous one is committing — but adds no tests.

At minimum I'd expect:
- An integration test verifying that auto-transactions actually fire (and that the `disable_auto_transaction` dev_tweak suppresses them).
- A test that ad-hoc query snapshots are consistent (not updated mid-transaction), since the snapshot-update guard at line 2953 is new and correctness-critical.
- A test covering the API-transaction-during-commit path (the `start_transaction_from_api` change at line 5423 that removed the `is_committing()` guard).

The existing test suite may exercise transactions in general, but these are novel edge cases that need explicit coverage.

File: crates/feldera-types/src/coordination.rs [hunk 2]
```diff
@@ -116,31 +116,46 @@ pub type Step = u64;
...
+    ///   execute at least one more step to commit it.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub transaction_status: Option<bool>,
 }
 
 impl StepStatus {
-    pub fn new(step: Step, action: StepAction) -> Self {
-        Self { step, action }
+    pub fn new(step: Step, action: StepAction, transaction_status: Option<bool>) -> Self {
+        Self {
```
Comment (mythical-fred) [inline crates/feldera-types/src/coordination.rs:148]: Soft: `Option<bool>` as a tri-state is a recurring pattern in the coordination types, but it's still a readability hazard — readers have to remember what `Some(true)` vs `Some(false)` vs `None` mean. A small enum (`TransactionStatus { None, Open, Committing }`) would be self-documenting and wouldn't cost anything at the wire level (serde can flatten it to the same JSON). Not blocking on this since the existing `TransactionCoordination` already used the same convention, but worth considering before the pattern spreads further.

**Review by mythical-fred — CHANGES_REQUESTED** [review]
Two findings, one hard blocker — see inline.

File: crates/feldera-types/src/coordination.rs (hunk 2 above)
Comment (blp) [reply crates/feldera-types/src/coordination.rs:148]: @ryzhyk I've thought this very same thing a couple of times myself. What has stopped me thus far is naming: we already have TransactionState, TransactionStatus, TransactionInfo, and probably more, and none is quite right. If there's a good name, or if we can rename things, then I'd prefer it over true/false.

I don't know whether it would be better as a tri-state or as an Option of Open vs. Committing.
